File: LeetcodePython/L152_MaximumProductSubarray.py
def maxProduct(self, nums: List[int]) -> int:

    # O(n) 
    # init res
    res = max(nums)

    # any numebr times 1 equal 1
    curMin,curMax = 1,1

    for n in nums:
        if n == 0:
            curMax,curMin = 1,1
            continue

        # maintain 2 values instead of dp
        tmp = n * curMax
        curMax = max(n * curMax, n * curMin,n)
        curMin = min(tmp, n * curMin,n)
        res = max(res,curMax)

    return res

File: 1_7/src/L152MaximumProductSubArray.java
import java.util.Arrays;

public class L152MaximumProductSubArray {
    public int maxProduct(int[] nums) {
        // use dp to solve
        int res = Arrays.stream(nums).max().getAsInt();

        int curMin = 1, curMax = 1;

        for ( int n : nums){
            if (n == 0){
                curMax = 1;
                curMin = 1;
                continue;
            }

            int tmp = curMax * n;
            curMax = Math.max(n,Math.max(n * curMax,n* curMin));
            curMin = Math.min(n,Math.min(tmp,n* curMin));

            res = Math.max(res,curMax);
        }

        return res;
    }
}
